File: android/app/src/main/java/de/x0bubbuff/needlebub/packs/PackManifest.kt
package de.x0bubbuff.needlebub.packs

import org.json.JSONException
import org.json.JSONObject

class PackValidationException(message: String, cause: Throwable? = null) : Exception(message, cause)

data class PackOutput(
    val type: String,
    val pointer: String,
    val optional: Boolean,
)

data class PackManifest(
    val formatVersion: Int,
    val id: String,
    val version: String,
    val name: String,
    val author: String,
    val description: String,
    val license: String,
    val engineAbi: String,
    val modelPath: String,
    val modelSize: Long,
    val modelSha256: String,
    val queryTemplate: String,
    val surfaces: Set<String>,
    val outputs: Map<String, PackOutput>,
) {
    companion object {
        const val FORMAT_VERSION = 1
        const val ENGINE_ABI = "needle2-hf-98fbd955b0347e78059be0c253cc1ffa09b87bc7-android-arm64"
        private val topLevelKeys = setOf(
            "formatVersion", "id", "version", "name", "author", "description", "license",
            "engine", "model", "queryTemplate", "surfaces", "outputs",
        )
        private val outputTypes = setOf("string", "boolean", "number", "json")
        private val allowedSurfaces = setOf("external", "notification")

        @Throws(PackValidationException::class)
        fun parse(raw: String): PackManifest {
            try {
                val json = JSONObject(raw)
                requireExactKeys(json, topLevelKeys, "manifest")
                if (json.getInt("formatVersion") != FORMAT_VERSION) invalid("unsupported formatVersion")

                val id = json.requiredString("id")
                if (id.length > 200) invalid("id is too long")
                if (!id.matches(Regex("^[a-z][a-z0-9]*(?:\\.[a-z0-9][a-z0-9-]*){2,}\$"))) invalid("id must use reverse-domain notation")
                val version = json.requiredString("version")
                if (version.length > 100) invalid("version is too long")
                if (!version.matches(Regex("^(0|[1-9]\\d*)\\.(0|[1-9]\\d*)\\.(0|[1-9]\\d*)(?:-[0-9A-Za-z.-]+)?(?:\\+[0-9A-Za-z.-]+)?\$"))) invalid("version must be semantic")

                val engine = json.getJSONObject("engine")
                requireExactKeys(engine, setOf("abi"), "engine")
                val engineAbi = engine.requiredString("abi")

                val model = json.getJSONObject("model")
                requireExactKeys(model, setOf("path", "size", "sha256"), "model")
                val modelPath = model.requiredString("path")
                if (modelPath != "model.cact") invalid("model.path must be model.cact")
                val modelSize = model.getLong("size")
                if (modelSize !in 1..MAX_MODEL_BYTES) invalid("model.size is out of range")
                val digest = model.requiredString("sha256")
                if (!digest.matches(Regex("^[a-fA-F0-9]{64}\$"))) invalid("model.sha256 must be SHA-256")

                val queryTemplate = json.requiredString("queryTemplate")
                if (queryTemplate.length > MAX_QUERY_TEMPLATE_CHARS) invalid("queryTemplate is too long")
                if (Regex(Regex.escape("{{input}}")).findAll(queryTemplate).count() != 1) invalid("queryTemplate needs exactly one {{input}}")

                val rawSurfaces = json.getJSONArray("surfaces")
                val surfaces = (0 until rawSurfaces.length()).map { rawSurfaces.getString(it) }.toSet()
                if (surfaces.isEmpty() || surfaces.size != rawSurfaces.length() || !allowedSurfaces.containsAll(surfaces)) invalid("invalid surfaces")

                val rawOutputs = json.getJSONObject("outputs")
                if (rawOutputs.length() > MAX_OUTPUTS) invalid("too many declared outputs")
                val outputs = rawOutputs.keys().asSequence().associateWith { name ->
                    if (!name.matches(Regex("^nb_[a-z][a-z0-9_]*\$"))) invalid("invalid output name")
                    val output = rawOutputs.getJSONObject(name)
                    requireExactKeys(output, setOf("type", "pointer", "optional"), "output $name")
                    val type = output.requiredString("type")
                    if (type !in outputTypes) invalid("invalid output type")
                    val pointer = output.requiredString("pointer")
                    if (!pointer.matches(Regex("^(?:/(?:[^~/]|~[01])*)*\$"))) invalid("invalid JSON Pointer")
                    PackOutput(type, pointer, output.optBoolean("optional", false))
                }

                return PackManifest(
                    FORMAT_VERSION, id, version, json.requiredString("name"), json.requiredString("author"),
                    json.requiredString("description"), json.requiredString("license"), engineAbi,
                    modelPath, modelSize, digest.lowercase(), queryTemplate, surfaces, outputs,
                )
            } catch (error: PackValidationException) {
                throw error
            } catch (error: JSONException) {
                throw PackValidationException("invalid manifest JSON", error)
            }
        }

        private fun requireExactKeys(json: JSONObject, allowed: Set<String>, label: String) {
            val unknown = json.keys().asSequence().filterNot(allowed::contains).toList()
            if (unknown.isNotEmpty()) invalid("$label contains unknown fields: ${unknown.joinToString()}")
        }

        private fun JSONObject.requiredString(name: String): String =
            getString(name).takeIf(String::isNotBlank) ?: invalid("$name is required")

        private fun invalid(message: String): Nothing = throw PackValidationException(message)

        private const val MAX_MODEL_BYTES = 128L * 1024L * 1024L
        private const val MAX_QUERY_TEMPLATE_CHARS = 8 * 1024
        private const val MAX_OUTPUTS = 32
    }
}

File: src/domain/packManifest.ts
export const PACK_FORMAT_VERSION = 1 as const

export type PackSurface = 'external' | 'notification'
export type OutputType = 'string' | 'boolean' | 'number' | 'json'

export type PackOutput = {
  type: OutputType
  pointer: string
  optional?: boolean
}

export type PackManifest = {
  formatVersion: typeof PACK_FORMAT_VERSION
  id: string
  version: string
  name: string
  author: string
  description: string
  license: string
  engine: { abi: string }
  model: { path: 'model.cact'; size: number; sha256: string }
  queryTemplate: string
  surfaces: PackSurface[]
  outputs: Record<string, PackOutput>
}

const TOP_LEVEL_KEYS = new Set([
  'formatVersion', 'id', 'version', 'name', 'author', 'description', 'license',
  'engine', 'model', 'queryTemplate', 'surfaces', 'outputs',
])
const OUTPUT_TYPES = new Set<OutputType>(['string', 'boolean', 'number', 'json'])
const SURFACES = new Set<PackSurface>(['external', 'notification'])

function object(value: unknown, label: string): Record<string, unknown> {
  if (!value || typeof value !== 'object' || Array.isArray(value)) {
    throw new Error(`${label} must be an object`)
  }
  return value as Record<string, unknown>
}

function exactKeys(value: Record<string, unknown>, allowed: Set<string>, label: string): void {
  const unknown = Object.keys(value).filter((key) => !allowed.has(key))
  if (unknown.length) throw new Error(`${label} contains unknown fields: ${unknown.join(', ')}`)
}

function nonEmpty(value: unknown, label: string): asserts value is string {
  if (typeof value !== 'string' || value.trim().length === 0) throw new Error(`${label} is required`)
}

export function validatePackManifest(value: unknown): PackManifest {
  const manifest = object(value, 'manifest')
  exactKeys(manifest, TOP_LEVEL_KEYS, 'manifest')

  if (manifest.formatVersion !== PACK_FORMAT_VERSION) throw new Error('unsupported pack format')
  if (typeof manifest.id !== 'string' || !/^[a-z][a-z0-9]*(?:\.[a-z0-9][a-z0-9-]*){2,}$/.test(manifest.id)) {
    throw new Error('id must be reverse-domain notation')
  }
  if (manifest.id.length > 200) throw new Error('id is too long')
  if (typeof manifest.version !== 'string' || !/^(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)(?:-[0-9A-Za-z.-]+)?(?:\+[0-9A-Za-z.-]+)?$/.test(manifest.version)) {
    throw new Error('version must be semantic')
  }
  if (manifest.version.length > 100) throw new Error('version is too long')
  for (const field of ['name', 'author', 'description', 'license'] as const) nonEmpty(manifest[field], field)

  const engine = object(manifest.engine, 'engine')
  exactKeys(engine, new Set(['abi']), 'engine')
  nonEmpty(engine.abi, 'engine.abi')

  const model = object(manifest.model, 'model')
  exactKeys(model, new Set(['path', 'size', 'sha256']), 'model')
  if (model.path !== 'model.cact') throw new Error('model path must be model.cact')
  if (!Number.isSafeInteger(model.size) || (model.size as number) <= 0 || (model.size as number) > 128 * 1024 * 1024) throw new Error('model size is out of range')
  if (typeof model.sha256 !== 'string' || !/^[a-f0-9]{64}$/i.test(model.sha256)) throw new Error('model digest must be SHA-256')

  nonEmpty(manifest.queryTemplate, 'queryTemplate')
  if (manifest.queryTemplate.length > 8 * 1024) throw new Error('queryTemplate is too long')
  if (manifest.queryTemplate.split('{{input}}').length - 1 !== 1) throw new Error('queryTemplate needs exactly one {{input}}')

  if (!Array.isArray(manifest.surfaces) || manifest.surfaces.length === 0 || new Set(manifest.surfaces).size !== manifest.surfaces.length ||
      manifest.surfaces.some((surface) => typeof surface !== 'string' || !SURFACES.has(surface as PackSurface))) {
    throw new Error('surfaces are invalid')
  }

  const outputs = object(manifest.outputs, 'outputs')
  if (Object.keys(outputs).length > 32) throw new Error('too many declared outputs')
  for (const [name, rawOutput] of Object.entries(outputs)) {
    if (!/^nb_[a-z][a-z0-9_]*$/.test(name)) throw new Error(`invalid output name: ${name}`)
    const output = object(rawOutput, `outputs.${name}`)
    exactKeys(output, new Set(['type', 'pointer', 'optional']), `outputs.${name}`)
    if (typeof output.type !== 'string' || !OUTPUT_TYPES.has(output.type as OutputType)) throw new Error(`invalid output type: ${name}`)
    if (typeof output.pointer !== 'string' || !/^(?:\/(?:[^~/]|~[01])*)*$/.test(output.pointer)) throw new Error(`invalid JSON Pointer: ${name}`)
    if (output.optional !== undefined && typeof output.optional !== 'boolean') throw new Error(`invalid optional flag: ${name}`)
  }

  return value as PackManifest
}
